### ISSUE TITLE
setlocale's lack of thread safety seems to cause crashes on Windows

### ISSUE THREAD

**Issue by hawicz** [json-c/json-c#409]
As mentioned later on in issue #407, calling json_tokener_parse from multiple threads, or possibly just multiple times, on Windows can result in a crash in setlocale().

It might be possible to work around this by calling 
```
_configthreadlocale( _ENABLE_PER_THREAD_LOCALE)
````
or a uselocale() implementation as described at:
https://stackoverflow.com/questions/6561723/thread-specific-locale-manipulation-in-c

or maybe we should just punt and say it's up to the caller to set the locale appropriately.

**Comment by dota17** [comment]
today, I took some time to reproduce this thread-crahed question but failed. this is a small probability event indeed. 
```Setlocale()``` is not a thread-safe function and will cause the program to crash when Multi-threaded setlocales change the locale message at the same time. As for solution, the best is using the ```_configthreadlocale( _ENABLE_PER_THREAD_LOCALE) ```to make ```setlocale()``` only work for the current thread, or maybe using the thread lock. Another way is to let the user care about the thread question and use the setlocale() before Multi-thread. So that we should delete the ```setlocale()``` in parse function and ```HAVE_SETLOCALE``` in complie file. In this respect, it is not a good way.
In fact, the biggest question is that we cannot reproduce this thread-crahed question. So I couldn't test different method.

**Comment by MrdUkk** [comment]
this is completely wrong in my opinion. json-c as a universal library should not operate on any "current OS locale". 
It should be locale-agnostic, because you can't control myriads of other systems passing JSON documents to yours one....
